velox/functions/prestosql/aggregates/PrestoHasher.cpp

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#define XXH_INLINE_ALL
+
+#include "velox/functions/prestosql/aggregates/PrestoHasher.h"
+#include "velox/external/xxhash.h"
+
+namespace facebook::velox::aggregate {
+
+namespace {
+
+int64_t rotateLeft(const int64_t& num, const int32_t& distance) {
+  auto unsignedShift = ((uint64_t)num) >> -distance;
+  return (num << distance) | unsignedShift;
+}
+
+template <typename T>
+FOLLY_ALWAYS_INLINE int64_t getHashForIntTypes(const T& value) {
+  return rotateLeft(value * 0xC2B2AE3D27D4EB4FL, 31) * 0x9E3779B185EBCA87L;
+}
+
+template <typename T>
+FOLLY_ALWAYS_INLINE int64_t
+getHashForIntTypes(const DecodedVector& vector, vector_size_t row) {
+  auto value = vector.valueAt<T>(row);
+  return getHashForIntTypes(value);
+}
+
+FOLLY_ALWAYS_INLINE int64_t
+getHashForVarCharTypes(const DecodedVector& vector, const vector_size_t row) {
+  auto input = vector.valueAt<StringView>(row);
+  int64_t hash = XXH64(input.data(), input.size(), 0);
+  return hash;
+}
+} // namespace
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::BIGINT>(vector_size_t row) {
+  return getHashForIntTypes<int64_t>(*_vector.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::INTEGER>(vector_size_t row) {
+  return getHashForIntTypes<int32_t>(*_vector.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::SMALLINT>(vector_size_t row) {
+  return getHashForIntTypes<int16_t>(*_vector.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::TINYINT>(vector_size_t row) {
+  return getHashForIntTypes<int8_t>(*_vector.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::DOUBLE>(vector_size_t row) {
+  using charPtr = unsigned char*;
+  auto value = (*_vector.get()).valueAt<double>(row);
+  const charPtr buf = reinterpret_cast<const charPtr>(&value);
+  int64_t u;
+  charPtr pu = reinterpret_cast<charPtr>(&u);
+  std::copy(buf, buf + sizeof(double), pu);
+  return getHashForIntTypes(u);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::DATE>(vector_size_t row) {
+  return getHashForIntTypes<int32_t>(*_vector.get(), row);
```

Comment:
Does this work because Date struct has a single int32_t field? This is hacky though.

Reply:
Yes , do you recommend another way in which I can get the raw int out for hash purposes ?

Reply:
I'd use `vector_->valueAt<Date>(row).days()`.